### PR TITLE
DDFFORM 6 linkskomponent liste

### PR DESCRIPTION
**Pull request by LasseStaus**
- Added paragraph-config for links paragraph.
- Changed structure of linkit suggestions
- Add paragraph--links.html & paragraph module.
- Added contrib module for easily deleting links.

#### Link to issue

[Jira](https://reload.atlassian.net/jira/software/c/projects/DDFFORM/boards/485?selectedIssue=DDFFORM-6)

[design PR](https://github.com/danskernesdigitalebibliotek/dpl-design-system/pull/347)

#### Description
Overall this PR adds a paragraph--links component for editors of the website to use. 
The paragraph is a single paragraph, where editors can add multiple links. 

Additionally I've added: 
 - paragraph module for handling paragraph preprocess logic
 - Preprocess on paragraphs for add a baseIconPath to be available in every paragraph to avoid large folder paths. 
 - Preprocess on paragraph--links to handle the logic for the links being added.  
 - Added contrib module for easily deleting links. I think it was unintuitive, that in order to delete an existing link. you had to erase all link/field text and then save. This provides a delete button. 
 - Small adjustedment in linkit suggestion. (This could be removed, since what i did is not "final" in any way.) 

#### How to test 

Test that you can add a paragraph with multiple links on an article
Test that the icon will change depending on wether the link provide is an/a: 

- Internal link (arrow icon)
- External link (external icon)
- Link to a PDF, both internal / external (file icon )
- Search query result at /search (search icon)
- Search query result at /advanced-search  (search icon)

#### Screenshot of the result

image

#### comments 

Currently there is no vertical spacing between paragraphs in general. 
Since we haven't decided on an implementation of this yet, i've not done anything about it. If we add it later in design-system it'll work anyway. 

#### questions

I am unsure about the general logic of my preprocess for handling preparation of the link attributes. Generally just comment on everything you don't understand or if something is unnecessary. 

I would like you to consider "security-aspects" in terms of handling files. Currently i am checking /allowing pdf|docx|txt. Comment on what you think about this approach. 

meta: Comment on this pr. Is the description sufficient, insufficient, overkill, anything else.

### CODE REVIEW

**Review by kasperg — CHANGES_REQUESTED** [review]
:+1: Nice stuff here.

In general the configuration looks right to me:

- I appreciate you adding the Multiple Fields Remove Button module. It seems to work well.
- We still need to take [our configuration management and the config ignore module into account](https://danskernesdigitalebibliotek.github.io/dpl-docs/dpl-cms/configuration-management/?h=configuration#install-a-new-site-from-scratch).

I enjoy the level of commenting in the PR but I would have preferred it to be a part of the commit comments instead of added separately. Most of what you have written under "Description" could have been commit comments.

Consider your level of commenting in the code. To be it gets too verbose. If a comment does not add value then leave it out. Personally I would like to see more type hints where they can not be automatically inferred so our editors and PHPStan can help us more.

Also note that when I add an link to an external component I get a warning in Lagoon. You probably want to look into that.

Monosnap Edit Article Voluptatibus eos hic ex nemo corporis | DPL CMS | Logged in 2023-12-03 15-49-19

Monosnap Voluptatibus eos hic ex nemo corporis | DPL CMS | Logget ind 2023-12-03 15-47-48

File: web/modules/custom/dpl_paragraphs/dpl_paragraphs.module [hunk 1]
```diff
@@ -0,0 +1,95 @@
+<?php
+
+/**
+ * @file
+ * DPL Base.
+ */
```
[the remark this answers is not in the document]
Comment (LasseStaus) [reply web/modules/custom/dpl_paragraphs/dpl_paragraphs.module:6]: I've added this to be ignored in phpcs. Hopefully i added it correctly, can be seen in the commit once pushed.

File: web/modules/custom/dpl_paragraphs/dpl_paragraphs.module [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      // Normalize the URL to extract only the path, excluding query parameters.
+      $normalized_url = rtrim(parse_url($url->toString(), PHP_URL_PATH), '/');
+
+      // Default link type is 'internal'.
+      $link_type = 'internal';
+
+      if ($url->isExternal()) {
+        // Check if the URL points to a downloadable file.
+        if (preg_match('/\.(' . $downloadable_file_extensions . ')(\?.*)?$/', $normalized_url)) {
+          $link_type = 'download';
```
[the remark this answers is not in the document]
Comment (LasseStaus) [reply web/modules/custom/dpl_paragraphs/dpl_paragraphs.module:60]: I have added a comment describe the approach, and its' limitations. 

I actually don't have any specifics on the security part. I think it was more in the sense of wether there was anything i should consider, if some users were to upload malicious files. 

Obviously all editors should be from libraries, which minimises the risk greatly. But i was just thinking wether there was anything "we do" in terms of further protecting from this, or it is handled by drupal already.

**Comment by LasseStaus** [comment]
Ift. din kommentar 
"Also note that when I add an link to an external component I get a warning in Lagoon. You probably want to look into that."

Så jeg har lagt et check ind når URL'en bliver normaliseret hvis den er null. Jeg tror det fikser det men det må du meget gerne tjekke efter. 

Jeg kan godt linke fra en side til en anden på lagoon nu. Er usikker på, om det var det, som man ikke kunne før. Jeg havde ikke selv tested på lagoon da jeg requestede første review, det husker jeg til næste.

**Review by kasperg — APPROVED** [review]
:+1: I appreciate the changes.

**Pull request merged**